velox/functions/sparksql/String.cpp

```diff
@@ -103,6 +103,286 @@ class Length : public exec::VectorFunction {
   }
 };
 
+class ConcatWs : public exec::VectorFunction {
+ public:
+  explicit ConcatWs(const std::optional<std::string>& separator)
+      : separator_(separator) {}
+
+  // Calculate the total number of bytes in the result.
+  size_t calculateTotalResultBytes(
+      const SelectivityVector& rows,
+      exec::EvalCtx& context,
+      std::vector<exec::LocalDecodedVector>& decodedArrays,
+      const std::vector<std::string>& constantStrings,
+      const std::vector<exec::LocalDecodedVector>& decodedStringArgs,
+      const exec::LocalDecodedVector& decodedSeparator) const {
+    size_t totalResultBytes = 0;
+    rows.applyToSelected([&](auto row) {
+      int32_t allElements = 0;
+      // Calculate size for array args.
+      for (int i = 0; i < decodedArrays.size(); i++) {
+        DecodedVector& arrayDecoded = *decodedArrays[i].get();
+        auto baseArray = arrayDecoded.base()->as<ArrayVector>();
+        auto rawSizes = baseArray->rawSizes();
+        auto rawOffsets = baseArray->rawOffsets();
+        auto indices = arrayDecoded.indices();
+        auto elements = baseArray->elements();
+        exec::LocalSelectivityVector nestedRows(context, elements->size());
+        nestedRows.get()->setAll();
+        exec::LocalDecodedVector elementsHolder(
+            context, *elements, *nestedRows.get());
+        auto& elementsDecoded = *elementsHolder.get();
```

Comment:
Perhaps, rearrange the code to avoid doing this per row and do that per batch of rows instead.